### DISCUSSION TITLE
Windows 2019 Installation with ipxe restart over and over

### DISCUSSION THREAD

**Discussion by Lucho1976** [ipxe/ipxe#1045] [Q&A] [answered]
Hello guys, how are you? I hope well. I am having a problem trying to install Windows 2019 with iPXE. I have dnsmasq configured as a DHCP Server with the settings below, and the boot.ipxe file as I transcribe below too. The problem I have, as I show you in the video attatched, is that once the Windows installation finishes and restarts, it boots again from the DHCP and everything is done over and over again. I've been looking for a way to solve it and haven't been able to find it. Would you know how to solve this? I appreciate it in advance.

**dnsmasq.d**
interface=enp0s3
bind-interfaces
domain=glampinghub.local
dhcp-range=192.168.0.181,192.168.0.190,255.255.255.0,12h
dhcp-option=option:router,192.168.0.1
dhcp-option=option:dns-server,1.1.1.1
dhcp-option=option:dns-server,8.8.8.8
enable-tftp
tftp-root=/pxeboot

#! boot config for BIOS systems
dhcp-match=set:bios-x86,option:client-arch,0
dhcp-boot=tag:bios-x86,firmware/ipxe.pxe

#! boot config for UEFI systems
dhcp-match=set:efi-x86_64,option:client-arch,7
dhcp-match=set:efi-x86_64,option:client-arch,9
dhcp-boot=tag:efi-x86_64,firmware/ipxe.efi

**boot.ipxe**
#!ipxe

set server_ip  192.168.0.180

kernel tftp://${server_ip}/images/win2019/wimboot/wimboot
initrd tftp://${server_ip}/images/win2019/boot/bcd bcd
initrd tftp://${server_ip}/images/win2019/boot/boot.sdi boot.sdi
initrd tftp://${server_ip}/images/win2019/sources/WinPE.wim boot.wim
boot

**bootconfig.ipxe**
#!ipxe
dhcp
chain tftp://192.168.1.180/config/boot.ixpe

[Grabación de pantalla desde 28-09-23 11_06_27.webm](https://github.com/ipxe/ipxe/assets/26300768/9c6654a6-f4e0-47e3-a4fa-3050d5222c85)

**Comment by NiKiZe** [comment, marked as answer]
Change bios/firmware to boot from disk first, PXE boot should be a temporary boot option.

**Reply by Lucho1976** [reply]
Thank you very much, NiKiZe, for your prompt response. I tested it in VirtualBox and it worked perfectly. At my job, an automated installation is being carried out, and I'm not sure if I have access to the BIOS to change the boot order. Do you know if this could be controlled from boot.ipxe or bootconfig.ipxe? That is (see below), somehow check if Windows is installed and, if so, ask it to continue with the disk boot, or the next item within the boot list (in case the network is configured first and then the hard drive). I am very grateful.


#!ipxe

isset ${os_installed} && goto os_installed

set server_ip 192.168.0.180
kernel tftp://${server_ip}/images/win2019/wimboot/wimboot
initrd tftp://${server_ip}/images/win2019/boot/bcd bcd
initrd tftp://${server_ip}/images/win2019/boot/boot.sdi boot.sdi
initrd tftp://${server_ip}/images/win2019/sources/WinPE.wim boot.wim
boot

:os_installed
echo Operating system is already installed, continue with hard disk boot
....

**Reply by NiKiZe** [reply]
You might be able to use exit or exit 1 from iPXE to boot next boot device, but there is no way to detect installed os.

If boot order is disk, PXE and there is no os on disk, then that will fail and boot PXE instead.

What you could have is a server side solution which logs successfully installed machines based on Mac or serial. If they where install you do exit, but this will require some serverside scripting. And I would personally avoid this in prod.

**Comment by Lucho1976** [comment]
Hi NiKiZe, thank you so much for your answer. I found this solution that works fine and I want to share with other users.

set server_ip  192.168.0.180

#! try boot from hard disk
sanboot --no-describe --drive 0x80 || goto install_windows

:install_windows
kernel tftp://${server_ip}/images/win2019/wimboot/wimboot
initrd tftp://${server_ip}/images/win2019/boot/bcd bcd
initrd tftp://${server_ip}/images/win2019/boot/boot.sdi boot.sdi
initrd tftp://${server_ip}/images/win2019/sources/WinPE.wim boot.wim
boot

**Reply by NiKiZe** [reply]
One think that is nice to know, with modern wimboot and standard winpe, you do not need BCD or boot.sdi.

Your solution only works in pcbios mode, not efi.
